R/list2df_linter.R

```diff
@@ -0,0 +1,64 @@
+#' Recommend direct usage of `data.frame()` to create a data.frame from a list
+#'
+#' [list2DF()] is the preferred way to turn a list of columns into a data.frame.
+#'   Note that it doesn't support recycling; if that's required, use [data.frame()].
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "do.call(cbind.data.frame, x)",
+#'   linters = list2df_linter()
+#' )
+#'
+#' lint(
+#'   text = "do.call('cbind.data.frame', x)",
+#'   linters = list2df_linter()
+#' )
+#'
+#' lint(
+#'   text = "do.call(cbind.data.frame, list(a = 1, b = 1:10))",
+#'   linters = list2df_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "list2df(x)",
+#'   linters = list2df_linter()
+#' )
+#'
+#' lint(
+#'   text = "data.frame(list(a = 1, b = 1:10))",
+#'   linters = list2df_linter()
+#' )
+#'
+#' @evalRd rd_tags("list2df_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+list2df_linter <- function() {
+
+  docall_nolambda_xpath <- "
+  //SYMBOL_FUNCTION_CALL[text() = 'do.call']
+    /parent::expr
+    /following-sibling::expr[1][SYMBOL or STR_CONST]
+  "
+
+  Linter(linter_level = "expression", function(source_expression) {
+    xml <- source_expression$xml_parsed_content
+    xml_calls <- xml_find_all(xml, docall_nolambda_xpath)
+    docall_fun <- get_r_string(xml_calls)
+
+    bad_expr <- xml_calls[
+      !is.na(docall_fun) & docall_fun == "cbind.data.frame"
```

Comment:
I think it's good to be safe, but now that you've landed on `SYMBOL or STR_CONST`, it should definitely not be `NA`, so it might be better to be consistent with other linters & omit the `is.na()` check.

The problem before was you did not include any condition on `SYMBOL or STR_CONST` in the initial XPath search, so `get_r_string()` returned NA for those nodes that didn't match. By including that condition up-front, we guarantee `get_r_string()` lands as intended.